### PR TITLE
Add a GError derive macro

### PR DESCRIPTION
**Pull request by elmarco**

### CODE REVIEW

File: glib-macros/src/gerror_derive.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use proc_macro2::TokenStream;
+use proc_macro_error::abort_call_site;
+use quote::{quote, quote_spanned};
+use syn::{punctuated::Punctuated, spanned::Spanned, token::Comma, Data, Ident, Variant};
+
+use crate::utils::{crate_ident_new, parse_domain_name};
+
+// FIXME: merge with genum version
```
Comment (sdroege) [inline glib-macros/src/gerror_derive.rs:10]: Is that difficult? If not, maybe just do it here in a separate commit :)

**Review by sdroege — COMMENTED** [review]
Looks mostly good to me, thanks!

File: glib-macros/src/lib.rs [hunk 2]
```diff
@@ -249,6 +250,31 @@ pub fn genum_derive(input: TokenStream) -> TokenStream {
...
+///
+/// # Example
+///
+/// ```
+/// use glib::prelude::*;
+/// use glib::subclass::prelude::*;
+///
+/// #[derive(Debug, Copy, Clone, glib::GError)]
+/// #[gerror(domain_name = "ExFoo")]
+/// enum Foo {
```
[the remark this answers is not in the document]
Comment (elmarco) [reply glib-macros/src/lib.rs:263]: why do you think it's necessary? It's the rust implementation after, doesn't have an influence on the ffi, right? Afaict, "self as i32" in code() is the only case where it would matter, but it should be fine whatever representation, isn't it?

File: glib-macros/src/lib.rs [hunk 3]
```diff
@@ -249,6 +250,31 @@ pub fn genum_derive(input: TokenStream) -> TokenStream {
...
+///     Blah,
+///     Baaz,
+/// }
+/// ```
+///
+/// [`ErrorDomain`]: error/trait.ErrorDomain.html
+#[proc_macro_derive(GError, attributes(gerror))]
```
[the remark this answers is not in the document]
Comment (elmarco) [reply glib-macros/src/lib.rs:270]: agree

File: glib-macros/src/gerror_derive.rs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            fn from(code: i32) -> Option<Self>
+            where
+                Self: Sized
+            {
+                #from_glib
+                None
+            }
+        }
+    }
+    // TODO: generate the extern "C" fn {}_quark(void) name.to_string().to_snake_case() ?
```
[the remark this answers is not in the document]
Comment (elmarco) [reply glib-macros/src/gerror_derive.rs:67]: Well, unfortunately, cbindgen doesn't generate the ffi, the crate has to provide it. Or am I missing something?

File: glib-macros/src/gerror_derive.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            "{}: derive(GError) requires #[gerror(domain_name = \"DomainName\")]",
+            e
+        ),
+    };
+    let from_glib = gen_enum_from_glib(name, enum_variants);
+
+    quote! {
+        impl #crate_ident::error::ErrorDomain for #name {
+            fn domain() -> #crate_ident::Quark {
+                #crate_ident::Quark::from_string(#domain_name)
```
[the remark this answers is not in the document]
Comment (elmarco) [reply glib-macros/src/gerror_derive.rs:51]: ok

File: glib-macros/src/gerror_derive.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (elmarco) [reply glib-macros/src/gerror_derive.rs:67]: However, one somewhat big issue if/when we start generating extern fn or other FFI from macros is that cbindgen doesn't see them... By default, it process unexpanded crates. If you expand crates and glib macros, then everything becomes also more complicated for gbindgen (as it rely on the macros...! the expanded version would be an extra layer of complexity)

What would be interesting would be to be able expand macros selectively. But at this point, this not possible (it's a rustc today limitation). Another option would be for gbindgen to process unexpanded & expanded version...?

File: glib-macros/src/gerror_derive.rs (hunk 1 above)
Comment (elmarco) [reply glib-macros/src/gerror_derive.rs]: done, hopefully in a correct way

File: glib-macros/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sdroege) [reply glib-macros/src/lib.rs:264]: Right, if you always go via `FromGlib` / `ToGlib` it's irrelevant indeed

File: glib-macros/src/gerror_derive.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sdroege) [reply glib-macros/src/gerror_derive.rs]: For now I would do like we always did: write the FFI functions manually. At some point we could consider having a code generator for that, I guess. Unlike macros that would then also not run into that problem.

But I don't think it should ever be the job of this derive macro to generate the FFI functions.

File: glib-macros/src/gerror_domain_derive.rs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            "{}: derive(GErrorDomain) requires #[gerror_domain(name = \"DomainName\")]",
+            e
+        ),
+    };
+    let from_glib = gen_enum_from_glib(name, enum_variants);
+
+    quote! {
+        impl #crate_ident::error::ErrorDomain for #name {
+            fn domain() -> #crate_ident::Quark {
+                #crate_ident::Quark::from_static_string(&concat!(#domain_name, "\0"))
```
Comment (sdroege) [inline glib-macros/src/gerror_domain_derive.rs:32]: There is no such function in the safe bindings (because of Rust not having a NUL-terminator in strings), you'll have to call FFI.

File: glib/src/quark.rs [hunk 7]
```diff
@@ -14,6 +14,11 @@ impl Quark {
         unsafe { from_glib(ffi::g_quark_from_string(s.to_glib_none().0)) }
     }
 
+    #[doc(alias = "g_quark_from_static_string")]
+    pub fn from_static_string(s: &'static str) -> Quark {
+        unsafe { from_glib(ffi::g_quark_from_static_string(s.to_glib_none().0)) }
```
Comment (sdroege) [inline glib/src/quark.rs:19]: This is unsound: the string is not actually static. `to_glib_none().0` has to allocate a new temporary string, append a NUL, pass it to the C function and then frees it.

That's why you directly have to call the FFI function

File: glib-macros/src/gerror_domain_derive.rs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+            fn from(value: i32) -> Option<Self>
+            where
+                Self: Sized
+            {
+                #from_glib
+            }
+        }
+    }
+    // TODO: generate the extern "C" fn {}_quark(void) name.to_string().to_snake_case() ?
```
Comment (sdroege) [inline glib-macros/src/gerror_domain_derive.rs:47]: (so remove this TODO comment, that's just confusing for now)

**Review by sdroege — COMMENTED** [review]
Can you also add a small test to glib-macros/tests, like the existing ones? Then we can at least be sure that it compiles and kind of behaves (e.g. you can create a `glib::Error` from it).

File: glib-macros/src/gerror_derive.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (elmarco) [reply glib-macros/src/gerror_derive.rs]: You are probably right.

File: glib/src/quark.rs (hunk 7 above)
Comment (elmarco) [reply glib/src/quark.rs:19]: ok

File: glib-macros/src/gerror_derive.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (elmarco) [reply glib-macros/src/gerror_derive.rs]: sorry, my change got lost.. doing it again

File: glib-macros/src/gerror_domain_derive.rs (hunk 6 above)
Comment (elmarco) [reply glib-macros/src/gerror_domain_derive.rs]: ah, that's why we use the once_cell then :)

Comment (sdroege) [reply glib-macros/src/gerror_domain_derive.rs]: No, the `Lazy` we use so we don't have to go through the global `GQuark` lock every time :)

**Comment by elmarco** [comment]
> Can you also add a small test to glib-macros/tests, like the existing ones? Then we can at least be sure that it compiles and kind of behaves (e.g. you can create a `glib::Error` from it).

done (it was partially covered by doc already)

File: glib-macros/src/genum_derive.rs [hunk 9]
```diff
@@ -112,10 +95,23 @@ pub fn impl_genum(input: &syn::DeriveInput) -> TokenStream {
             }
         }
 
+        impl #crate_ident::translate::TryFromGlib<i32> for #name {
```
Comment (sdroege) [inline glib-macros/src/genum_derive.rs:98]: Why is this needed btw?

**Review by sdroege — APPROVED** [review]

File: glib-macros/src/genum_derive.rs (hunk 9 above)
Comment (elmarco) [reply glib-macros/src/genum_derive.rs:98]: ErrorDomain::from() -> Option, so it was simpler for the common code generator to also return Option. I could have not implemented TryFromGlib and just inlined the call and unwrap() in FromGlib, but I thought that could also be useful (more useful that FromGlib that may panic)

**Review by GuillaumeGomez — APPROVED** [review]
